multicluster/controllers/multicluster/node_controller.go

```diff
@@ -0,0 +1,202 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: r.namespace,
+		},
+	}
+	// When the Node is annotated with 'multicluster.antrea.io/gateway=true' as a Gateway:
+	//   - Delete the Gateway if the Node is deleted
+	//   - Update the Gateway if Node's InternalIP or GatewayIP is updated
+	//   - Create a new Gateway if there is no existing Gateway
+	node := &corev1.Node{}
+	if err := r.Client.Get(ctx, req.NamespacedName, node); err != nil {
+		if apierrors.IsNotFound(err) {
+			if err := r.Client.Delete(ctx, gwNode, &client.DeleteOptions{}); err != nil {
+				return ctrl.Result{}, client.IgnoreNotFound(err)
+			}
+			return ctrl.Result{}, nil
+		}
+		klog.ErrorS(err, "Failed to get Node", "node", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	_, isGWNode := node.Annotations[common.GatewayAnnotation]
+	var err error
+	var gwIP, internalIP string
+	if internalIP, gwIP, err = r.getGatawayNodeIP(node); err != nil {
+		return ctrl.Result{}, err
+	}
+	gwNodeNamespaced := types.NamespacedName{
+		Name:      node.Name,
+		Namespace: r.namespace,
+	}
+
+	if err := r.Client.Get(ctx, gwNodeNamespaced, gwNode); err != nil {
+		if apierrors.IsNotFound(err) && isGWNode {
+			gwNode.GatewayIP = gwIP
+			gwNode.InternalIP = internalIP
+			if err := r.Client.Create(ctx, gwNode, &client.CreateOptions{}); err != nil {
+				return ctrl.Result{}, err
+			}
+			r.installedGWNodes.Add(gwNode)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if !isGWNode {
+		if err := r.Client.Delete(ctx, gwNode, &client.DeleteOptions{}); err != nil {
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	gwNode.GatewayIP = gwIP
+	gwNode.InternalIP = internalIP
+	if err := r.updateGateway(node, gwNode); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *NodeReconciler) updateGateway(node *corev1.Node, gwNode *mcsv1alpha1.Gateway) error {
+	cached, exist, _ := r.installedGWNodes.GetByKey(node.Name)
+	if exist {
+		cachedGWNode := cached.(*mcsv1alpha1.Gateway)
+		if cachedGWNode.GatewayIP == gwNode.GatewayIP && cachedGWNode.InternalIP == gwNode.InternalIP {
+			klog.Info("No change for the Gateway, skip updating", "gateway", gwNode.Name)
+			return nil
+		}
+	}
+	if err := r.Client.Update(ctx, gwNode, &client.UpdateOptions{}); err != nil {
+		return err
+	}
+	r.installedGWNodes.Update(gwNode)
+	return nil
+}
+
+func (r *NodeReconciler) getGatawayNodeIP(node *corev1.Node) (string, string, error) {
+	var gatewayIP, internalIP string
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == corev1.NodeInternalIP {
+			if r.precedence == mcsv1alpha1.PrecedencePrivate {
+				gatewayIP = addr.Address
+			}
+			internalIP = addr.Address
+		}
+		if r.precedence == mcsv1alpha1.PrecedencePublic && addr.Type == corev1.NodeExternalIP {
+			gatewayIP = addr.Address
+		}
+	}
+
+	if ip, ok := node.Annotations[common.GatewayIPAnnotation]; ok {
+		gatewayIP = ip
```

Comment:
We should validate the value is an IP, and if not log error and ignore the Node.

Reply:
fixed.